### PR TITLE
FEATURE: BB-849 Previous period measure definition support

### PR DESCRIPTION
**Pull request by martinmosko**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/590/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/169/ : SUCCESS in 5m 23s

**Comment by martinmosko** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/80/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/591/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/37/ : SUCCESS in 17m 29s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/593/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/170/ : SUCCESS in 1m 49s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/594/ : SUCCESS in 14s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/171/ : SUCCESS in 1m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/595/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/172/ : SUCCESS in 1m 41s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/598/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/175/ : SUCCESS in 2m 17s

File: src/factory/DerivedMeasureTitleSuffixFactory.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import IntlStore from '../helpers/IntlStore';
+import IMeasureDefinitionType = VisualizationObject.IMeasureDefinitionType;
+
+/**
+ * Factory that builds formatted localized suffix string for derived measure based on the measure definition type.
+ * The suffix is used during AFM execution and for bucket item titles.
+ *
+ * @internal
+ */
+export default class DerivedMeasureTitleSuffixFactory {
```
[the remark this answers is not in the document]
Comment (martinmosko) [reply src/factory/DerivedMeasureTitleSuffixFactory.ts]: As we had a discussion with @tomasreichmann. Suggested solution was not fully satisfying for us. But he allowed us to keep it even he was not convinced about advantage for using class as factory.

File: src/helpers/popHelper.ts [hunk 2]
```diff
@@ -29,67 +25,71 @@ function getAllMeasureBucketItems(
...
+        return definition.popMeasureDefinition.measureIdentifier;
+    } else if (VisualizationObject.isPreviousPeriodMeasureDefinition(definition)) {
+        return definition.previousPeriodMeasure.measureIdentifier;
+    } else {
+        return null;
     }
+}
+
+function getMasterMeasure(bucketItems: IMeasure[], measureIdentifier: string): IMeasure {
+    return bucketItems.find(bucketItem => get(bucketItem, ['measure', 'localIdentifier']) === measureIdentifier);
```
[the remark this answers is not in the document]
Comment (martinmosko) [reply src/helpers/popHelper.ts]: fixed :)

File: src/helpers/popHelper.ts [hunk 3]
```diff
@@ -29,67 +25,71 @@ function getAllMeasureBucketItems(
...
-                ? masterMeasureTitle
-                : masterMeasureAlias;
+            if (masterMeasureIdentifier === null) {
+                return bucketItem;
+            }
 
-            const titleProp = { title: `${derivedMeasureTitleBase}${popSuffix}` };
+            const masterMeasure = getMasterMeasure(measureBucketItems, masterMeasureIdentifier);
+            const derivedMeasureTitleBase = getDerivedMeasureTitleBase(masterMeasure);
+            const title = `${derivedMeasureTitleBase}${suffixFactory.getSuffix(definition)}`;
```
[the remark this answers is not in the document]
Comment (martinmosko) [reply src/helpers/popHelper.ts]: good point :)

[the remark this answers is not in the document]
Comment (martinmosko) [reply src/helpers/popHelper.ts]: fixed

File: src/helpers/popHelper.ts [hunk 4]
```diff
@@ -29,67 +25,71 @@ function getAllMeasureBucketItems(
...
+    } else {
+        return null;
     }
+}
+
+function getMasterMeasure(bucketItems: IMeasure[], measureIdentifier: string): IMeasure {
+    return bucketItems.find(bucketItem => get(bucketItem, ['measure', 'localIdentifier']) === measureIdentifier);
+}
+
+function getDerivedMeasureTitleBase(masterMeasure: VisualizationObject.IMeasure): string {
```
[the remark this answers is not in the document]
Comment (martinmosko) [reply src/helpers/popHelper.ts]:
```

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/610/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/184/ : SUCCESS in 1m 43s

**Comment by tomasreichmann** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/611/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/185/ : SUCCESS in 7m 33s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/616/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/197/ : SUCCESS in 4m 18s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/97/artifact/test-storybook/html-report/index.html

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/96/artifact/test-storybook/html-report/index.html

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/98/artifact/test-storybook/html-report/index.html

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/99/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/617/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/8/ : SUCCESS in 25m 19s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/618/ : SUCCESS in 0s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/199/ : FAILURE in 11m 49s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/621/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/201/ : SUCCESS in 2m 13s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/120/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/202/ : SUCCESS in 7m 57s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/100/ : SUCCESS in 18m 09s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/8/ : SUCCESS in 11m 56s

**Pull request merged**